firefox_puppeteer/ui/about_window/window.py

```diff
@@ -0,0 +1,142 @@
+
+    def download(self, wait_for_finish=True, timeout=TIMEOUT_UPDATE_DOWNLOAD):
+        """ Download the update.
+
+        :param wait_for_finish: Optional, if True the function has to wait
+        for the download to be finished, default to `True`
+        :param timeout: Optional, How long to wait for the download to finish,
+        default to 360 seconds
+        """
+        assert self._software_update.update_channel.default_channel == \
+            self._software_update.update_channel.channel, \
+            'The update channel has been set correctly. ' \
+            'default_channel: is {}, while channel is: {}'.format(
+                self._software_update.update_channel.default_channel,
+                self._software_update.update_channel.channel)
+
+        if self.deck.selected_panel == self.deck.download_and_install:
+            self.deck.download_and_install.button.click()
+
+            # Wait for the download to start
+            Wait(self.marionette).until(
+                lambda _: self.deck.selected_panel != self.deck.download_and_install)
+
+        # If there are incompatible addons we fallback on old software update dialog for updating
+        if self.deck.selected_panel == self.deck.apply_billboard:
+            # The rest of the code inside this method uses the update wizard which is
+            # not being converted yet, so raise a NotImplementedError.
+            raise NotImplementedError('Fallback dialog logic not yet implemented.')
+
+            # self.deck.update_button.click()
+            #
+            # The current JS code is:
+            #
+            #     var wizard = updateWizard.handleUpdateWizardDialog();
+            #     wizard.waitForWizardPage(updateWizard.WIZARD_PAGES.updatesfoundbasic);
+            #     wizard.download();
+            #     wizard.close();
+            #     this._downloadDuration = wizard._downloadDuration;
+            #
+            #     return;
+
+        if wait_for_finish:
+            start_time = datetime.now()
+            self.wait_for_download_finished(timeout)
+            self._download_duration = datetime.now() - start_time
+
+    def wait_for_check_finished(self, timeout=TIMEOUT_UPDATE_CHECK):
+        """Waits for update checking to complete.
+
+        :param timeout: Optional, How long to wait for the update check to finish,
+        default to 30 seconds
+        """
+        Wait(self.marionette, self.TIMEOUT_UPDATE_CHECK).until(
+            lambda _: self.deck.selected_panel != self.deck.check_for_updates and
+            self.deck.selected_panel != self.deck.checking_for_updates,
+            message='An update has been found.')
+
+    def wait_for_download_finished(self, timeout=TIMEOUT_UPDATE_DOWNLOAD):
+        """ Waits until download is completed.
+
+        :param timeout: Optional, How long to wait for the download to finish,
+        default to 360 seconds
+        """
+        Wait(self.marionette, timeout).until(
```

Comment:
@whimboo This does not seem to be working. The code does not wait here, but just continues. I added the extra condition that the selected panel is not the `download_and_install` panel, in addition to not being the `downloading` panel, but it doesn't seem to be working.